### PR TITLE
Add spec: deserializing objects with parameterized ctors

### PR DESCRIPTION
**Pull request by layomia**
Contributes to https://github.com/dotnet/runtime/issues/29895.

### CODE REVIEW

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 1]
```diff
@@ -0,0 +1,900 @@
...
+
+string serialized = JsonSerializer.Serialize(new Point(1, 2), options);
+Console.WriteLine(serialized); // {"x_value":1,"y_value":2}
+
+Point point = JsonSerializer.Deserialize<Point>(serialized, options);
+Console.WriteLine(point.XValue); // 1
+Console.WriteLine(point.YValue); // 2
+```
+
+An alternative solution is to share the `PropertyNamingPolicy` used by object properties.
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Here's some conventions I think make sense:
- A ctor argument typically maps 1:1 with a property. This is required for round-tripping the JSON\object. That property likely only has a getter (no setter), but having a setter shouldn't affect this feature since the ctor will still be called.
- A ctor argument is the name of the corresponding property but camel-cased (the property was Pascal-cased).

So consider an approach where we do not treat the object's properties and ctor args as two independent things and that the property and argument naming policies work together.

For example for Serialize it works the same as today:
- The "XValue" properties is serialized as "x_value" based on a snake-casing property naming policy.

And for deserialize it would:
1) Assume the "x_value" JSON is matched to the "XValue" property (already done today).
2) In order to determine if a ctor should be called, the new "argument naming policy" is run (default is camel-case) and converts "XValue" (the property name) to "xValue".
3) The "xValue" is found in the ctor arg (case-sensitive) and then that ctor is called (assuming "yValue" will also be handled like this).

This is nice because it will support `JsonNameAttribute`. If we had a `JsonName["HelloWorld"]` on the "XValue" property, the JSON would be "HelloWorld" (since it ignores the "property naming policy") which would map back to the "XValue" property (like today) which would then be converted to "xValue" argument by the "argument naming policy".

To simplify this, we could just assume the "argument naming policy" is always the camel-casing policy and not expose a way to change that on JsonSerializerOptions. If some edge cases don't work (because camel-casing is not used for argument names) then the workaround is to create a custom converter or create a new type of custom converter that will make this a bit easier (not yet public\designed but we want it for other reasons).

[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: For cases when (1) above does not hold because the corresponding property isn't found from the JSON -- then on deserialize we could:

Run the "property naming policy" on the ctor argument names like we do today for properties. Thus we treat argument as "pseudo-properties". Thus "xValue" ctor arg produces "x_value". This would be cached and not done on every deserialization.

For deserialize:
1) The "x_value" JSON did not match up to any properties but did match up to the ctor arg "xValue".
2) Run the "argument naming policy" against "xValue" (which produces "xValue" since it was already camel-cased). So when "property naming policy" is snake-case or camel-case this doesn't really do anything, but would work perfect when no "property naming policy" is specified and we have Pascal-cased JSON.
3) The "x_value" JSON is matched to the "xValue" arg and that ctor is called...

For simplicity we could just do this for all cases (don't try to match up argument name and property name) -- it just means the `JsonName["HelloWorld"]` scenario wouldn't work.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 2]
```diff
@@ -0,0 +1,900 @@
...
+```
+
+It might be non-trivial work to resolve such scenarios on deserialization and handle error cases.
+
+This feature work will not include honoring references within objects deserialized with parameterized constructors. It may be considered in the future.
+
+`Nestonsoft.Json` does not not honor reference metadata within objects deserialized with
+parameterized constructors.
+
+### For performance, properties that may be deserialized through constructor arguments will be written first on deserialization.
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Yes, this approach brings some good optimizations spanning various scenarios. Here are some numbers for simple classes with  with a mix of `class` and `struct` primitives:

Method | Mean | Error | StdDev | Median | Min | Max | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Deserialize_Parameterless | 2.956 us | 0.0362 us | 0.0339 us | 2.945 us | 2.917 us | 3.015 us | 1.00 | 0.00 | 0.0710 | - | - | 336 B
Deserialize_Parameterized_LessArgs_CtorArgsFirstJsonOrder | 3.344 us | 0.0584 us | 0.0517 us | 3.328 us | 3.283 us | 3.475 us | 1.13 | 0.02 | 0.0801 | - | - | 376 B
Deserialize_Parameterized_MoreArgs_CtorArgsFirstJsonOrder | 3.504 us | 0.0435 us | 0.0407 us | 3.491 us | 3.454 us | 3.571 us | 1.19 | 0.02 | 0.1095 | - | - | 504 B
Deserialize_Parameterized_LessArgs_AverageJsonOrder | 4.114 us | 0.0455 us | 0.0426 us | 4.117 us | 4.042 us | 4.167 us | 1.39 | 0.03 | 0.0808 | - | - | 376 B
Deserialize_Parameterized_MoreArgs_AverageJsonOrder | 4.367 us | 0.0684 us | 0.0639 us | 4.361 us | 4.259 us | 4.487 us | 1.48 | 0.03 | 0.1043 | - | - | 504 B

These are the types being deserialized:

```C#
public class Parameterless_ClassWithPrimitives
{
    public int FirstInt { get; set; }
    public int SecondInt { get; set; }
    public string FirstString { get; set; }
    public string SecondString { get; set; }
    public DateTime FirstDateTime { get; set; }
    public DateTime SecondDateTime { get; set; }
    public int X { get; set;  }
    public int Y { get; set;  }
    public int Z { get; set; }
    public int ThirdInt { get; set; }
    public int FourthInt { get; set; }
    public string ThirdString { get; set; }
    public string FourthString { get; set; }
    public DateTime ThirdDateTime { get; set; }
    public DateTime FourthDateTime { get; set; }
}

public class Parameterized_ClassWithPrimitives_3Args
{
    public int FirstInt { get; set; }
    public int SecondInt { get; set; }
    public string FirstString { get; set; }
    public string SecondString { get; set; }
    public DateTime FirstDateTime { get; set; }
    public DateTime SecondDateTime { get; set; }
    public int X { get; }
    public int Y { get; }
    public int Z { get; }
    public int ThirdInt { get; set; }
    public int FourthInt { get; set; }
    public string ThirdString { get; set; }
    public string FourthString { get; set; }
    public DateTime ThirdDateTime { get; set; }
    public DateTime FourthDateTime { get; set; }

    public Parameterized_ClassWithPrimitives_3Args(int x, int y, int z) => (X, Y, Z) = (x, y, z);
}

public class Parameterized_ClassWithPrimitives_8Args
{
    public int FirstInt { get; set; }
    public int SecondInt { get; set; }
    public string FirstString { get; set; }
    public string SecondString { get; set; }
    public DateTime FirstDateTime { get; }
    public DateTime SecondDateTime { get; }
    public int X { get; }
    public int Y { get; }
    public int Z { get; }
    public int ThirdInt { get; }
    public int FourthInt { get; }
    public string ThirdString { get; }
    public string FourthString { get; set; }
    public DateTime ThirdDateTime { get; set; }
    public DateTime FourthDateTime { get; set; }


    public Parameterized_ClassWithPrimitives_8Args(
        DateTime firstDateTime,
        DateTime secondDateTime,
        int x,
        int y,
        int z,
        int thirdInt,
        int fourthInt,
        string thirdString)
    {
        FirstDateTime = firstDateTime;
        SecondDateTime = secondDateTime;
        X = x;
        Y = y;
        Z = z;
        ThirdInt = thirdInt;
        FourthInt = fourthInt;
        ThirdString = thirdString;
    }
}
```

I'll provide some more details on the numbers and scenarios in an oncoming PR in dotnet/performance.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: I like the proposal to link the property names to the constructor parameter names, as it allows roundtripping scenarios where a `[JsonPropertyName]` is applied to a property, as you bring up.
If there's no property matching the constructor arg, the `PropertyNamingPolicy` in the options will used to determine what JSON to look for.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 3]
```diff
@@ -0,0 +1,900 @@
...
+        /// The default value is false.
+        /// </summary>
+        /// <remarks>There is a performance cost associated when the value is true.</remarks>
+        public bool ParameterNameCaseInsensitive { get { throw null; } set { } }
+
+        /// <summary>
+        /// Specifies the policy used to convert a parameter's name on an object to another format, such as camel-casing.
+        /// The resulting parameter name is expected to match the JSON payload during deserialization.
+        /// </summary>
+        public System.Text.Json.JsonNamingPolicy? ParameterNamingPolicy { get { throw null; } set { } }
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: The initial proposal was to add this as a way to specify what JSON property name the parameter should look for.

https://github.com/dotnet/runtime/pull/33095/commits/a913b8088fbcef612d0b7a86e8ff4e42fbe41493 introduces an updated proposal where constructor parameters

- match with object properties, and use their cached metadata instead
- fallback on the `PropertyNamingPolicy` if there's no matching property.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 4]
```diff
@@ -0,0 +1,900 @@
...
+
+    [JsonConstructor]
+    public Point(int x, int y) => (X, Y) = (x, y);
+}
+```
+
+We can deserialize JSON into an instance of `Point` using `JsonSerializer`:
+
+```C#
+Point point = JsonSerializer.Deserialize<Point>(@"{""X"":1,""Y"":2}");
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Yes the policy applied on object properties to match with constructor arguments is `camelCase` by default. This proposal doesn't include a way to change this for now, but should work for most scenarios, including ASP.NET. We can expose an option on the future.

Yup, ASP.NET scenarios are support "OOB", but they might have to add `[JsonConstructor]` where applicable.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 5]
```diff
@@ -0,0 +1,900 @@
...
+    
+    - More flexible.
+    - Unnecessary perf hit in the major scenario (`camelCase` parameters, `PascalCase` properties).
+
+2. Do case insensitive matching and provide an opt-in for case sensitive matching.
+
+    - More flexible.
+    - Unnecessary perf hit in the major scenario if option is not specified.
+
+3. Instead of adding the `ParameterCaseInsensitive` option, share the `PropertyCaseInsensitive` option
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Yes, I agree that they would probably always be the same casing (properties and constructor parametes), so it makes sense for the same option to control them.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: I update the matching semantics as follows:

- Object properties are bound to constructor parameters by converting the properties using  `camelCase` as described in https://github.com/dotnet/runtime/pull/33095#discussion_r387844875. Constructor parameters will then take over the cached JSON property names from the object properties (which will be ignored on deserialization)
- If there's no match for a constructor parameter, the policy specified with `options.PropertyNamingPolicy` will be used to determine what JSON property names to match with.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: More benchmarks here: https://gist.github.com/layomia/9e7a67e7c89d5658df037409757965c6.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 6]
```diff
@@ -0,0 +1,834 @@
...
+This is an implementation detail. If deserialization is attempted with a constructor that has more than
+64 parameters, a `NotSupportedException` will be thrown.
+
+We expect most users to have significantly less than 64 parameters, but we can respond to user feedback.
+
+#### [`ReferenceHandling` semantics](https://github.com/dotnet/runtime/blob/13c1e65a9f7aab201fe77e3daba11946aeb7cbaa/src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md) will not be applied to objects deserialized with parameterized constructors
+
+`NotSupportedException` will be thrown if any properties named "$id", "$ref", or "$values" are found in the payload, and `options.ReferenceHandling` is set to 
+`ReferenceHandling.Preserve`. If the feature is off, the these properties will be treated like any other. This behavior prevents us from breaking people if we
+implement this feature in the future.
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]:
```suggestion
`ReferenceHandling.Preserve`. If the feature is off, these properties will be treated like any other. This behavior prevents us from breaking people if we implement this feature in the future.
```

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 7]
```diff
@@ -0,0 +1,834 @@
...
+
+`Newtonsoft.Json` does not not honor reference metadata within objects deserialized with parameterized constructors.
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: It just ignores them?

**Review by jozkee — APPROVED** [review]
Otherwise; looks great.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 8]
```diff
@@ -0,0 +1,834 @@
...
+Console.WriteLine(json); // {"XValue":1,"YValue":2}
+
+point = JsonSerializer.Deserialize<Point>(json);
+Console.WriteLine(point.X);
+Console.WriteLine(point.Y);
+```
+
+**If a constructor parameter does not match with a property, then the `JsonNamingPolicy` specified in `options.PropertyNamingPolicy` will be used to assign a JSON property name**. The default value for `options.PropertyNamingPolicy` is `null`, so no naming policy is applied by default.
+
+**Parameter naming matching is case-sensitive by default**. This can be toggled by users with the `options.PropertyNameCaseInsensitive` option.
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: It is probably not needed as there won't be a lot of situations where they won't be the same value.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md (hunk 7 above)
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Yup - I'll mention this explicitly.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 9]
```diff
@@ -0,0 +1,862 @@
+# Deserializing objects using parameterized constructors with `JsonSerializer`
+
+## Motivation
+
+`JsonSerializer` deserializes instances of objects (`class`es and `struct`s) using public parameterless
+constructors. If none is present, and deserialization is attempted, the serializer throws a `NotSupportedException`
+with a message stating that objects without public parameterless constructors, including `interface`s and `abstract`
+types, are not supported for deserialization. There is no way to deserialize an instance of an object using a parameterized constructor.
+
+A common pattern is to make data objects immutable for various reasons. For example, given `Point`:
```
Comment (terrajobst) [inline src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Please note that this change alone won't support all immutable types. For example, our own immutable collections don't use constructors but factory methods.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 10]
```diff
@@ -0,0 +1,862 @@
...
+public class Point
+{
+    public int X { get; }
+
+    public int Y { get; }
+
+    public int Z { get; }
+
+    [JsonConstructor]
+    private Point(int x, int y) => (X, Y) = (x, y);
```
Comment (terrajobst) [inline src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: I'm a bit concerned about features like this because most people (and that includes us) don't really reason about privates as external state. I'd sleep better if we could make this `protected`, for instance.

Comment (steveharter) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Agree. We should verify that we really, really need non-public support.

`private` and `protected` support won't support most code-gen designs -- the converter and POCO must be one and the same. `internal` is a little more palatable.

Comment (terrajobst) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Depends on how we do the source gen. Wasn't one proposal that we'd put them on the same type?

Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: > We should verify that we really, really need non-public support

Agreed. We have a similar concern with the upcoming non-public getter/setter support (https://github.com/dotnet/runtime/issues/29743) feature, but we are committed to doing this.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Can you share a TL;DR/summary of the benchmarks and what's the conclusion?

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+{
+  public int X { get; }
+
+  public int Y { get; }
+
+  public Point(int x, int y) => (X, Y) = (x, y);
+}
+```
+
+The `int X` property matches with the `int x` parameter, and the `int Y` property matches with the `int x` property.
```
Comment (campersau) [inline src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]:
```suggestion
The `int X` property matches with the `int x` parameter, and the `int Y` property matches with the `int y` parameter.
```

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md (hunk 10 above)
Comment (ericstj) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: I think if folks realize what they are doing it might be ok.  Something like this:
image

Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: The resolution is to wait for an explicit customer ask before adding this.

File: src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: Benchmarks: https://github.com/dotnet/performance/pull/1233.

TLDR

- There's a ~4% regression for existing benchmarks when deserializing type graphs with numerous parameterless ctors (specifically `IndexViewModel` and `MyEventListerViewModel` benchmarks). There's also a ~300B increase in allocs for these cases. This is due to new state added to the `ReadStack` to support this feature, and a minimal check when pushing/popping stack frames to support fetching ctor arg state on continuation and/or scenarios involving nested objects with parameterized ctors.
- There's a minimum alloc cost of 200B when deserializing objects with parameterized ctors, compared to deserializing the same object using parameteless ctors (baseline), assuming all members are deserialized in both cases. Total alloc increase depends on the number of ctor args, and how many of them are value types.
- Deserialization of "immutable" objects (where every JSON property maps with a ctor arg) with parameterized ctors is 0-15% slower than baseline. Smaller object graphs are slower due to the overhead of setting up a cache for the arguments.
- Deserialization of mutable objects (where some JSON properties map to object members) with is 0-67% slower than baseline. The slowest cases are when there are numerous object members to deserialize. This is due to re-reading JSON properties after ctor args are parsed and the object is constructed.

[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md]: FWI - following spec review and feedback that we should have "last constructor arguments wins" semantics to stay consistent with the behavior for properties, this approach of writing the ctor args first is no longer beneficial.
